pkg/database/database.go

```diff
@@ -31,19 +44,37 @@ type DBInterface interface {
 	QueryNamespacedCoreResources(ctx context.Context, kind, version, namespace string) ([]*unstructured.Unstructured, error)
 	WriteResource(ctx context.Context, k8sObj *unstructured.Unstructured, data []byte) error
 	Ping(ctx context.Context) error
+	TestConnection(env map[string]string) error
 }
 
 type Database struct {
-	db                *sql.DB
-	resourceTableName string
+	db   *sql.DB
+	info DBInfoInterface
 }
 
-func NewDatabase() (*Database, error) {
-	dataSource, err := ConnectionStr()
+func NewDatabase() (DBInterface, error) {
+	env, err := NewDatabaseEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	var database DBInterface
+	if f, ok := RegisteredDatabases[env[DbKindEnvVar]]; ok {
+		database = f()
+	} else {
+		panic(fmt.Sprintf("No database registered with name %s", env[DbKindEnvVar]))
+	}
+
+	err = database.TestConnection(env)
 	if err != nil {
 		return nil, err
 	}
-	var db *sql.DB
+	return database, nil
+}
+
+func (db *Database) TestConnection(env map[string]string) error {
```

Comment:
I think that this method name is missleading, it states `TestConnection` but if I understand correctly this is doing more than that (the method that just test the connection is the `Ping()`), this is establishing the database connection inside the `Database` object. So here are several things to consider:
* This method shouldn't be called from the outside (api/sink), it's just meant to be used from the database constructor so I believe it shouldn't be part of the `DBInterface`. We are falling into the [interface pollution](https://100go.co/5-interface-pollution/) practice.
* When the `NewDatabase` actually creates the database object by running the proper function of the `RegisteredDatabases` map it returns and non-functional database because the db connection hasn't been established. I would suggest to establish the connection in the `New<db>Database` instead of here.
* As  setting the connection is common code to be used, I think it makes sense to extract it to another file and provide it as a "tool" for the different database implementations to use. it would just have to receive the driver and the connection string and return the connection object `sql.DB`. Please take a look at the `connection.go` file in https://github.com/kubearchive/kubearchive/pull/461 where I do that

Reply:
Fixed.